package org.devopology.common.converter;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class DoubleConverterTest {

    @Test
    public void testValidConversion1() throws ConverterException {
        Object object = "12345.0";
        Double expected = Double.parseDouble((String) object);
        Double actual = Converter.DOUBLE.convert(object);

        assertEquals(expected, actual);
    }

    @Test
    public void testValidConversion2() throws ConverterException {
        Object object = "12345";
        Double expected = Double.parseDouble((String) object);
        Double actual = Converter.DOUBLE.convert(object);

        assertEquals(expected, actual);
    }

    @Test
    public void testInvalidConversion1() throws ConverterException {
        Assertions.assertThrows(IllegalArgumentException.class, () -> {
            Converter.DOUBLE.convert(null);
        });
    }

    @Test
    public void testInvalidConversion2() throws ConverterException {
        Assertions.assertThrows(ConverterException.class, () -> {
            Converter.DOUBLE.convert("");
        });
    }

    @Test
    public void testInvalidConversion3() throws ConverterException {
        Assertions.assertThrows(ConverterException.class, () -> {
            Converter.DOUBLE.convert(" ");
        });
    }

    @Test
    public void testInvalidConversion4() throws ConverterException {
        Assertions.assertThrows(ConverterException.class, () -> {
            Converter.DOUBLE.convert("12345x");
        });
    }

    @Test
    public void testInvalidConversion5() throws ConverterException {
        Assertions.assertThrows(ConverterException.class, () -> {
            Converter.DOUBLE.convert("x12345");
        });
    }

    @Test
    public void testInvalidConversion6() throws ConverterException {
        Assertions.assertThrows(ConverterException.class, () -> {
            Converter.DOUBLE.convert(new Object());
        });
    }
}
